File: javascript/src/lambda_expression.js
// FIXME: This is a work in progress. The code is not yet complete.

// Define States
const Water = (state) => state.Water || false;
const BoiledWater = (state) => state.BoiledWater || false;
const CoffeeBeans = (state) => state.CoffeeBeans || false;
const GroundCoffee = (state) => state.GroundCoffee || false;
const Filter = (state) => state.Filter || false;
const CoffeeInPot = (state) => state.CoffeeInPot || false;
const Pending = (state) => state.Pending || false;
const Sent = (state) => state.Sent || false;
const Payment = (state) => state.Payment || false;
const Cup = (state) => state.Cup || false;

// Transitions as Lambda Functions
const BoilWater = (state) =>
  Water(state) && !BoiledWater(state)
    ? (s) => ({
        ...s,
        Water: false,
        BoiledWater: true,
      })
    : (s) => s;

const GrindBeans = (state) =>
  CoffeeBeans(state) && !GroundCoffee(state)
    ? (s) => ({
        ...s,
        CoffeeBeans: false,
        GroundCoffee: true,
      })
    : (s) => s;

const BrewCoffee = (state) =>
  BoiledWater(state) && GroundCoffee(state) && Filter(state) && !CoffeeInPot(state)
    ? (s) => ({
        ...s,
        BoiledWater: false,
        GroundCoffee: false,
        Filter: false,
        CoffeeInPot: true,
      })
    : (s) => s;

const PourCoffee = (state) =>
  CoffeeInPot(state) && Cup(state)
    ? (s) => ({
        ...s,
        CoffeeInPot: false,
        Cup: false,
        Pending: true,
      })
    : (s) => s;

const Send = (state) =>
  Pending(state)
    ? (s) => ({
        ...s,
        Pending: false,
        Sent: true,
      })
    : (s) => s;

const Credit = (state) =>
  Sent(state)
    ? (s) => ({
        ...s,
        Sent: false,
        Payment: true,
      })
    : (s) => s;

// Guards as Lambda Functions
const Guard = (state, transition) => {
  const invalid = (s) => {
    // Example guard logic (e.g., you cannot PourCoffee without Payment)
    if (transition === PourCoffee && !Payment(s)) return true;
    return false;
  };
  return !invalid(state);
};

// Execute Transition (Higher-Order Function)
const executeTransition = (state, transition) =>
  Guard(state, transition) ? transition(state)(state) : state;

// Initial State
const initialState = {
  Water: true,
  CoffeeBeans: true,
  Filter: true,
  Cup: true,
  Pending: false,
  BoiledWater: false,
  GroundCoffee: false,
  CoffeeInPot: false,
  Sent: false,
  Payment: false,
};

// Sequence of Transitions (Lambda Composition)
const processCoffeeMachine = (state) => {
  const transitions = [BoilWater, GrindBeans, BrewCoffee, PourCoffee, Send, Credit];
  return transitions.reduce((currentState, transition) => {
    console.log("Current State:", currentState);
    const newState = executeTransition(currentState, transition);
    console.log("New State:", newState);
    return newState;
  }, state);
};

// Run the Coffee Machine
const finalState = processCoffeeMachine(initialState);
console.log("Final State:", finalState);

File: python/lambda_expression.py
from enum import Enum, auto

class State(Enum):
    WATER = auto()
    BOILED_WATER = auto()
    COFFEE_BEANS = auto()
    GROUND_COFFEE = auto()
    FILTER = auto()
    COFFEE_IN_POT = auto()
    PENDING = auto()
    SENT = auto()
    PAYMENT = auto()
    CUP = auto()

# CoffeeMachineState is a dictionary of States to bool
CoffeeMachineState = dict

# Transition Functions
def boil_water(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.WATER) and not state.get(State.BOILED_WATER):
        state[State.BOILED_WATER] = True
        state[State.WATER] = False
    return state

def grind_coffee(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.COFFEE_BEANS) and not state.get(State.GROUND_COFFEE):
        state[State.GROUND_COFFEE] = True
        state[State.COFFEE_BEANS] = False
    return state

def brew_coffee(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.BOILED_WATER) and state.get(State.GROUND_COFFEE) and state.get(State.FILTER):
        state[State.COFFEE_IN_POT] = True
        state[State.BOILED_WATER] = False
        state[State.GROUND_COFFEE] = False
        state[State.FILTER] = False
    return state

def send_order(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.PENDING) and not state.get(State.SENT):
        state[State.SENT] = True
        state[State.PENDING] = False
    return state

def make_payment(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.SENT) and not state.get(State.PAYMENT):
        state[State.PAYMENT] = True
        state[State.SENT] = False
    return state

def serve_coffee(state: CoffeeMachineState) -> CoffeeMachineState:
    if state.get(State.COFFEE_IN_POT) and state.get(State.PAYMENT) and not state.get(State.CUP):
        state[State.CUP] = True
        state[State.COFFEE_IN_POT] = False
        state[State.PAYMENT] = False
    return state

# Example usage
if __name__ == "__main__":
    state = {
        State.WATER: True,
        State.BOILED_WATER: False,
        State.COFFEE_BEANS: False,
        State.GROUND_COFFEE: False,
        State.FILTER: False,
        State.COFFEE_IN_POT: False,
        State.PENDING: False,
        State.SENT: False,
        State.PAYMENT: False,
        State.CUP: False,
    }
    state = boil_water(state)
    state = grind_coffee(state)
    state = brew_coffee(state)
    state = send_order(state)
    state = make_payment(state)
    state = serve_coffee(state)
    print(state)